README.md
# gunicorn_worker_healthcheck

HTTP endpoint that counts ready gunicorn workers

When using the default `sync` gunicorn worker, using an HTTP endpoint of the application itself
(i.e. that is also handled by the same mechanism, namely the gunicorn workers, as normal application traffic)
to check whether the application is able to handle traffic
(and to e.g. remove it from the load balancing rotation on failure)
can lead to negative feedback loops in high load situations:

The workers may become so backed up with application requests that they are unable to respond to the healthcheck requests in a timely manner.
Thus, this application is removed from serving traffic, 
leaving even less applications to serve the incoming traffic,
thus increasing the chance that they, too, will become backed up 
-- repeat until no applications are available for serving traffic at all.

This package provides an alternative healthcheck endpoint that 
a) is served by a separate thread, independent of the main application traffic
b) only checks whether the application and gunicorn startup process has completed
(i.e. any/all gunicorn worker processes have reported to be ready).


## Usage

Put this into your config file (usually `gunicorn.conf.py`):

```
import gunicorn_worker_healthcheck

healthcheck_bind = '127.0.0.1:8001'
gunicorn_worker_healthcheck.start(globals())
```

This starts an HTTP server in a thread; 
on request it counts the number of workers that have notified they are ready
(using the [`post_worker_init` hook](https://docs.gunicorn.org/en/latest/settings.html#post-worker-init) to write a pid file).
If the required number of workers are ready, it returns HTTP 200, else 500.


## Configuration

* `healthcheck_bind` host:port on which to listen for healthcheck requests (default: None, i.e. healthcheck is disabled)
* `healthcheck_require_workers` count of workers that must be ready for the healthcheck to report success (default: 1)
* `healthcheck_directory` directory in which to store the state files, one per worker (default: `''`, i.e. the current directory)
* `healthcheck_filename` format string how to name the state files, must contain one placeholder `{}` for the pid (default: `'.gunicorn.worker.{}'`)


CHANGES.md
## gunicorn_worker_healthcheck Changes

<!-- towncrier release notes start -->

## 1.0.0 (2023-05-19)





- Initial release (initial)


src/gunicorn_worker_healthcheck/test/fixture/app.py
def wsgi(environ, start_response):
    start_response('200 OK', [('Content-type', 'text/plain; charset=ascii')])
    return ['OK\n'.encode('ascii')]


src/gunicorn_worker_healthcheck/test/fixture/conf.py
bind = '127.0.0.1:{main}'
healthcheck_bind = '127.0.0.1:{check}'

workers = 2

import gunicorn_worker_healthcheck
gunicorn_worker_healthcheck.start(globals())


src/gunicorn_worker_healthcheck/test/test_healthcheck.py
from time import sleep
import contextlib
import importlib.resources
import pytest
import requests
import socket
import subprocess
import sys


def random_port():
    s = socket.socket()
    with contextlib.closing(s):
        s.bind(('localhost', 0))
        return s.getsockname()[1]


@pytest.fixture()
def server(tmp_path):
    main = random_port()
    check = random_port()

    config = importlib.resources.files(__package__).joinpath(
        'fixture/conf.py').open('r').read()
    (tmp_path / 'gunicorn.conf.py').write_text(
        config.format(main=main, check=check))
    proc = subprocess.Popen(
        [sys.executable, '-m', 'gunicorn',
         'gunicorn_worker_healthcheck.test.fixture.app:wsgi'],
        cwd=str(tmp_path))
    yield {'main': main, 'check': check}
    proc.terminate()
    proc.wait(5)


def test_runs_healthcheck_on_separate_port(server):
    http = requests.Session()
    timeout = 10
    for _ in range(timeout):
        sleep(1)
        r = http.get('http://localhost:%s' % server['check'])
        if r.status_code == 200:
            assert r.text.strip() == '2 workers ready'
            r = http.get('http://localhost:%s' % server['main'])
            assert r.status_code == 200
            assert r.text.strip() == 'OK'
            break
    else:
        pytest.fail(f'Did not start inside {timeout} seconds')


src/gunicorn_worker_healthcheck/__init__.py
from glob import glob
from wsgiref.simple_server import WSGIServer, WSGIRequestHandler
import os
import logging
import threading


__version__ = '1.1.0.dev0'


def start(cfg):
    bind = cfg.get('healthcheck_bind')
    if not bind:
        return

    check = HealthCheck(
        bind,
        cfg.get('healthcheck_directory', ''),
        cfg.get('healthcheck_filename', '.gunicorn.worker.{}'),
        cfg.get('healthcheck_require_workers', 1))
    check.register(cfg)
    check.start()


class HealthCheck(threading.Thread):

    def __init__(self, bind, directory, filename, workers):
        self.bind = bind
        self.directory = directory
        self.filename = filename
        self.workers = workers

        WSGIRequestHandler.log_request = lambda *args: None  # sigh
        host, port = self.bind.split(':')
        self.server = WSGIServer((host, int(port)), WSGIRequestHandler)
        self.server.set_app(self.wsgi_app)

        super().__init__(daemon=True, target=self.server.serve_forever)

    def register(self, cfg):
        cfg['when_ready'] = self.log_port
        cfg['post_worker_init'] = self.add_worker
        cfg['worker_exit'] = self.remove_worker
        cfg['on_exit'] = self.stop

    def log_port(self, server):
        logging.getLogger('gunicorn.error').info(
            'Healthcheck listening at: http://%s', self.bind)

    def add_worker(self, worker):
        open(self.pidfile(worker.pid), 'w').close()

    def remove_worker(self, worker, server):
        f = self.pidfile(worker.pid)
        if os.path.exists(f):
            os.remove(f)

    def pidfile(self, name):
        return os.path.join(self.directory, self.filename.format(name))

    def stop(self, server):
        for f in glob(self.pidfile('*')):
            os.remove(f)

        self.server.shutdown()
        self.server.server_close()
        self.join(1)

    def wsgi_app(self, environ, start_response):
        count = len(list(glob(self.pidfile('*'))))
        if count >= self.workers:
            status = '200 OK'
        else:
            status = '500 Internal Server Error'
        start_response(status, [('Content-type', 'text/plain; charset=ascii')])
        return [f'{count} workers ready\n'.encode('ascii')]
